### ISSUE TITLE
My trained model detect 0 detetctions

### ISSUE THREAD

**Issue by hasantha-nirmal** [berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset#3]
I have trained and followed the exact procedure. But when I run a test image there are no detetcions
![chart](https://user-images.githubusercontent.com/57673710/105030580-00270800-5a7a-11eb-8a08-e281bca82212.png)

**Comment by berktepebag** [comment]
Hello,
I'm a bit far from home now and will be back in ~2 weeks. I will try to check but I have no access to a GPU here, so it would be hard to replicate the error for me. But some questions:

1- What GPU you are using? (This was trained on 1070 8Gb lesser memory can be the reason)
2- Which OS? Windows or Linux?
3- Have you checked if the images are in one folder and they are correctly addressed? (If you are using Windows / and \ are opposite of Linux in the pathname, that can be the reason)
4- I have provided a trained model (weights under backup), have you tried that?  If it is working we can focus on the training process.

**Comment by hasantha-nirmal** [comment]
Thank you so much for the reply. I will answer my best to tell the situation

1.I have 2 GB NVIDIA-GEFORCE MX130 Graphic Card
2.I'm running on Ubuntu 18.04 LTS
3.Yes, I have checked and all the images are in 'traffic_light_images' folder as you commanded.And I have the final 'traffic_lights.txt' in darnet/traffic-lights folder.I will attach an snap with that.
[traffic_lights.txt](https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/files/5836741/traffic_lights.txt)

Yes, your model runs and detect them for some extent(for my test data anyway).I'm sorry I didn't know about your trained model before :(. My apologies.

Thank you again for reaching me so quickly. Please ask any question or request any material in your demand. I will provide them ASAP.Thanks again for your help.

Some files you might find useful
These files are in darknet/traffic-light folder
[bosch_traffic_light_xmls_list.txt](https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/files/5836762/bosch_traffic_light_xmls_list.txt)

[test.txt](https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/files/5836765/test.txt)
[train.txt](https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/files/5836770/train.txt)

I attach these by simply renaming these as txt files for convience of uploading.

[voc-bosch-data.txt](https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/files/5836783/voc-bosch-data.txt)
[voc-bosch-names.txt](https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/files/5836784/voc-bosch-names.txt)
[yolov3-tiny-bosch-cfg.txt](https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/files/5836785/yolov3-tiny-bosch-cfg.txt)


Please note that I have changed some parameters in yolov3-tiny-cosch-cfg file that you havaen't mention like max batch size.
But even before when I try to train with your exact information in this without changing any of those, I have got no detection at all with those iteration backups like 17600 th iteration backup

**Comment by berktepebag** [comment]
Well folders seems ok, my concern is 2Gb ram. Try to reduce the training batch (yolov3-tiny-bosch.cf 6th line) as low as 8-16 so we can be sure that they are fitting into the ram. Maybe 64 is causing a crash.

**Comment by hasantha-nirmal** [comment]
Thank you. I will surely try that and let you know. Please note that I have change the batch size and steps in yolov3-bosh.cfg too. I hope it won’t be a problem too ;)

**Comment by sandeepshrigondekar** [comment]
Hi - try to use -thresh during detection. This helps minimize the number of FNs (i.e., the true traffic lights that are erroneously left undetected).

**Comment by hasantha-nirmal** [comment]
> Hi - try to use -thresh during detection. This helps minimize the number of FNs (i.e., the true traffic lights that are erroneously left undetected).

Thanks for the help. I did that too. I even set it for 0 for any kind of hope. But unfortunately there was no detection whatsoever.

**Comment by berktepebag** [comment]
Have you tried the provided weights @hasantha-nirmal ? 

By the way thank you @sandeepshrigondekar, I will also give it a try when I'm back at home. Do the provided weights work for you?

**Comment by sandeepshrigondekar** [comment]
> 
> 
> Have you tried the provided weights @hasantha-nirmal ?
> 
> By the way thank you @sandeepshrigondekar, I will also give it a try when I'm back at home. Do the provided weights work for you?

Yes provided weights worked but now I am more interested in getting metrics on continuation to the approach. Any help on that would be greatly appreciated.

**Comment by hasantha-nirmal** [comment]
> Have you tried the provided weights @hasantha-nirmal ?
> 
> By the way thank you @sandeepshrigondekar, I will also give it a try when I'm back at home. Do the provided weights work for you?

Yes, provided weights worked for me. But I can’t train my own :(

**Comment by sandeepshrigondekar** [comment]
> 
> 
> > Have you tried the provided weights @hasantha-nirmal ?
> > By the way thank you @sandeepshrigondekar, I will also give it a try when I'm back at home. Do the provided weights work for you?
> 
> Yes, provided weights worked for me. But I can’t train my own :(

@hasantha-nirmal - how did you get the graph with this dataset that you posted with this issue ? Have you been able to get any metrics? If yes can you help on what all did you do

**Comment by hasantha-nirmal** [comment]
> > > Have you tried the provided weights @hasantha-nirmal ?
> > > By the way thank you @sandeepshrigondekar, I will also give it a try when I'm back at home. Do the provided weights work for you?
> > 
> > 
> > Yes, provided weights worked for me. But I can’t train my own :(
> 
> @hasantha-nirmal - how did you get the graph with this dataset that you posted with this issue ? Have you been able to get any metrics? If yes can you help on what all did you do

Everything I did , attached herewith.Please check them if you will.Yes I trained them but no matter how many times I did, I get no detections with my weights.

**Comment by hasantha-nirmal** [comment]
> > > > Have you tried the provided weights @hasantha-nirmal ?
> > > > By the way thank you @sandeepshrigondekar, I will also give it a try when I'm back at home. Do the provided weights work for you?
> > > 
> > > 
> > > Yes, provided weights worked for me. But I can’t train my own :(
> > 
> > 
> > @hasantha-nirmal - how did you get the graph with this dataset that you posted with this issue ? Have you been able to get any metrics? If yes can you help on what all did you do
> 
> Everything I did , attached herewith.Please check them if you will.Yes I trained them but no matter how many times I did, I get no detections with my weights.

https://github.com/berktepebag/Traffic-light-detection-with-YOLOv3-BOSCH-traffic-light-dataset/issues/3#issuecomment-762914020

**Comment by berktepebag** [comment]
> > > > Have you tried the provided weights @hasantha-nirmal ?
> > > > By the way thank you @sandeepshrigondekar, I will also give it a try when I'm back at home. Do the provided weights work for you?
> > > 
> > > 
> > > Yes, provided weights worked for me. But I can’t train my own :(
> > 
> > 
> > @hasantha-nirmal - how did you get the graph with this dataset that you posted with this issue ? Have you been able to get any metrics? If yes can you help on what all did you do
> 
> Everything I did , attached herewith.Please check them if you will.Yes I trained them but no matter how many times I did, I get no detections with my weights.

I would like to remind that it took almost 30-45 minutes with 1070 and 64 batch size. Smaller batches means that it will need more time to train (most of the time). You can check it [here](https://stats.stackexchange.com/questions/164876/what-is-the-trade-off-between-batch-size-and-number-of-iterations-to-train-a-neu). Model training is bit of a tweaking the knobs, so you should keep trying till you find a good solution, for example try to increase the iterations (1070=>64x40000 == mx 150=> 16x80000). It is hard to see what is wrong without trying, but I remember I had to try many differnt parameters till it started finding the traffic lights.

**Comment by hasantha-nirmal** [comment]
Thanks ; I will try my best and let you know sir

**Comment by joker-wutong** [comment]
I also have this problem, have you solved it?

**Comment by berktepebag** [comment]
> I also have this problem, have you solved it?

Can you share your system? I will do a walkthrough soon,  it may be useful.

**Comment by joker-wutong** [comment]
**Thank you so much for the reply. These pictures can illustrate my situation very well** @berktepebag 


![Screenshot from 2021-08-01 20-34-37](https://user-images.githubusercontent.com/61103650/127810543-adc2eff8-376f-4d0a-a655-94badd1adece.png)

![chart_yolov3-voc-bosch](https://user-images.githubusercontent.com/61103650/127810569-727029cf-64ad-4779-a97f-586912cd56f1.png)

**So what should I do, I hope to hear from you！**

**Comment by berktepebag** [comment]
I am not sure if Xavier (even though it is a good edge computing unit) is good for training anything at all. It is around 1050 gpu so problem can be the same. Can you try reducing batch size and increasing the iterations.

**Comment by joker-wutong** [comment]
I would like to ask whether there is a trained weight file for download？
thanks! @berktepebag

**Comment by berktepebag** [comment]
backup/yolov3-tiny-bosch_40000.weights can be used for detection purposes.

**Comment by alekhyaduba** [comment]
I was facing the same issue however I tried it https://github.com/AlexeyAB/darknet framework and with the same weights file I could get the predictions.
